### PR TITLE
Fix control lost focus on multi touch

### LINKED ISSUE #16761: Touch release event is not being executed on _gui_event method
**Godot version:**
Godot v3.0-stable

**OS/device including version:**
Huawei Honor 5X KIW-L21
Android 6.0.1 Marshmallow

**Issue description:**
The method `_gui_event(event)` is not being executed when the **last touch** (_InputEventScreenTouch_) is released. At least it doesn't work on Android platform, the Windows version with emulated touch events works properly.

**Steps to reproduce:**
Using one finger
- Touch the screen with your finger
- Method `_gui_event(event)` with `event.pressed = true` executed
- Release the finger
- Should execute `_gui_event(event)` with `event.pressed = false` touch event, but it don't

Multitouch doesn't work as well, but it lost only the last release event
- Touch the screen with your finger
- Touch the screen with your second finger
- Release your first finger
- Works ok, `_gui_event(event)` executed with `event.pressed = false` and `event.index == 0`
- Release your second finger
- Should execute `_gui_event(event)` with `event.pressed = false` and `event.index == 1` touch event, but it don't

**Minimal reproduction project:**
[gui-input-issue.zip](https://github.com/godotengine/godot/files/1732753/gui-input-issue.zip)

### PR DESCRIPTION
**Pull request by lethiandev**
Fixes https://github.com/godotengine/godot/issues/16761

Allows focusing multiple control elements on multi-touch screen devices. Also sets valid pointer's index 
 on release event when the pointers are released in reverse order.

Tested on 3.0 and 3.1, works on my KIW-L21, Android 6.0

Example project provided [_gui_input.zip](https://github.com/godotengine/godot/files/2503143/_gui_input.zip)

EDIT: I've made this PR because i'm working on a customized version of Godot 3.0 for my game and i've fixed this issue as well, maybe you can use it too.

### CODE REVIEW

File: scene/main/viewport.h [hunk 1]
```diff
@@ -251,6 +251,7 @@ class Viewport : public Node {
 		bool key_event_accepted;
 		Control *mouse_focus;
 		Control *mouse_click_grabber;
+		Vector<Control*> touch_focus;
```
Comment (LikeLakers2) [inline scene/main/viewport.h]:
```suggestion
		Vector<Control *> touch_focus;
```

**Comment by lethiandev** [comment]
It's fixed now, thanks.

**Comment by lethiandev** [comment]
I did more in-game tests and I've found that drag events cause crash on empty screen. now it's fixed too.

**Comment by akien-mga** [comment]
@RandomShaper Is this something you could help review?

**Comment by RandomShaper** [comment]
Yeah, I'll have a look.

**Comment by reduz** [comment]
Sorry, but I am against adding multitouch support for UI stuff. It's entirely not designed for this, and not a common use case. I believe whathever you may need is a corner case and should be handled as such.

**Comment by lethiandev** [comment]
I don't mind if my solution is not suitable for merge, but maybe information provided can be still useful to fix the problem with _gui_input. The problem occurs on events ordering when a `gui.mouse_focus` variable is set to NULL on InputEventMouseButton release event ([Line 1755](https://github.com/godotengine/godot/blob/master/scene/main/viewport.cpp#L1755)) and then it comes to handle a next incoming event InputEventScreenTouch, which results in no response.

**Comment by RandomShaper** [comment]
@lethiandev , thanks for taking it constructively!

Let me add to @reduz's response that changing the UI to multi-pointer has multipe implications that should be assessed very carefully. If it's ever requested by enough users for common use cases at any point, we can discuss its feasibility for an upcoming major upgrade (because that is compatibility-breaking).

That said, currently having a single-pointer UI system is a design decision that makes sense.

Anyway, I'll check your code for inspiration to fix the issue and credit you accordingly. Anyway, you are still invited to submit a new pull request addressing just that.

**Comment by lethiandev** [comment]
@RandomShaper Ok, tomorrow I'll make another PR addressing only this issue, thanks.
